### ISSUE TITLE
[趋势共振量化策略] | Trend Resonance Quant

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4387]
**标签**: quantitative-strategy, trend-following, technical-indicators, momentum
### 摘要
结合均线系统、MACD、RSI多指标共振确认趋势强度，过滤噪音信号，构建高胜率趋势跟踪策略。
### 算法描述
趋势跟踪是量化交易最经典的方向之一。多指标共振通过多个独立指标从不同角度验证趋势，提高信号的可靠性。

**均线系统（MA）**：
- SMA(n) = Σᵢ Pᵢ / n
- EMA(n) = α × Pᵢ + (1-α) × EMA_{i-1}，α = 2/(n+1)
- 金叉：短期MA上穿长期MA → 做多信号
- 死叉：短期MA下穿长期MA → 做空信号

**MACD 指标**：
- DIF = EMA(12) - EMA(26)
- DEA = EMA(DIF, 9)
- MACD Histogram = 2 × (DIF - DEA)
- DIF > 0 且 DIF > DEA：多头趋势
- DIF < 0 且 DIF < DEA：空头趋势

**RSI 指标**：
- RS = Σ Gain / Σ Loss（n日内）
- RSI = 100 - 100/(1 + RS)
- RSI > 70：超买区域
- RSI < 30：超卖区域

**共振评分系统**：
Score = w1×MA_signal + w2×MACD_signal + w3×RSI_signal
当 Score ≥ 阈值时触发信号，权重根据各指标历史胜率动态调整。

**趋势强度度量**：
- ADX > 25：趋势明确
- ADX < 20：震荡市，应降低仓位或退出

### Python代码示例
```python
import numpy as np
import pandas as pd
from dataclasses import dataclass, field
from typing import Dict, List, Tuple, Optional
from datetime import datetime, timedelta

@dataclass
class TrendSignal:
    """趋势信号"""
    timestamp: datetime
    direction: int  # 1: long, -1: short, 0: flat
    strength: float  # 0-1
    indicators: Dict[str, float]
    confidence: float

class MovingAverageCalculator:
    """均线计算器"""
    
    def __init__(self, short_window: int = 10, long_window: int = 30):
        self.short = short_window
        self.long = long_window
    
    def sma(self, prices: np.ndarray, period: int) -> np.ndarray:
        """简单移动平均"""
        result = np.full_like(prices, np.nan)
        for i in range(period - 1, len(prices)):
            result[i] = np.mean(prices[i - period + 1:i + 1])
        return result
    
    def ema(self, prices: np.ndarray, period: int) -> np.ndarray:
        """指数移动平均"""
        result = np.full_like(prices, np.nan)
        alpha = 2 / (period + 1)
        result[0] = prices[0]
        for i in range(1, len(prices)):
            result[i] = alpha * prices[i] + (1 - alpha) * result[i - 1]
        return result
    
    def calculate_ma_signals(self, prices: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
        """计算均线交叉信号"""
        short_ma = self.ema(prices, self.short)
        long_ma = self.ema(prices, self.long)
        signal = np.zeros_like(prices)
        for i in range(1, len(prices)):
            if np.isnan(short_ma[i]) or np.isnan(long_ma[i]):
                continue
            if short_ma[i] > long_ma[i] and short_ma[i-1] <= long_ma[i-1]:
                signal[i] = 1  # 金叉
            elif short_ma[i] < long_ma[i] and short_ma[i-1] >= long_ma[i-1]:
                signal[i] = -1  # 死叉
            elif short_ma[i] > long_ma[i]:
                signal[i] = 0.5  # 多头排列
            elif short_ma[i] < long_ma[i]:
                signal[i] = -0.5  # 空头排列
        return signal, short_ma
    
    def calculate_ma_strength(self, prices: np.ndarray) -> float:
        """计算均线趋势强度"""
        if len(prices) < self.long:
            return 0.0
        short_ma = self.ema(prices, self.short)
        long_ma = self.ema(prices, self.long)
        if np.isnan(short_ma[-1]) or np.isnan(long_ma[-1]):
            return 0.0
        price_position = (prices[-1] - short_ma[-1]) / (short_ma[-1] + 1e-10)
        ma_position = (short_ma[-1] - long_ma[-1]) / (long_ma[-1] + 1e-10)
        return np.clip((price_position + ma_position) / 2, -1, 1)

class MACDCalculator:
    """MACD 计算器"""
    
    def __init__(self, fast: int = 12, slow: int = 26, signal: int = 9):
        self.fast = fast
        self.slow = slow
        self.signal = signal
    
    def calculate(self, prices: np.ndarray) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
        """计算 MACD 系列"""
        ema_fast = MovingAverageCalculator(0, 0).ema(prices, self.fast)
        ema_slow = MovingAverageCalculator(0, 0).ema(prices, self.slow)
        dif = ema_fast - ema_slow
        dea = MovingAverageCalculator(0, 0).ema(dif, self.signal)
        histogram = 2 * (dif - dea)
        return dif, dea, histogram
    
    def calculate_macd_signals(self, prices: np.ndarray) -> np.ndarray:
        """MACD 交易信号"""
        dif, dea, hist = self.calculate(prices)
        signal = np.zeros_like(prices)
        for i in range(1, len(prices)):
            if np.isnan(dif[i]) or np.isnan(dea[i]):
                continue
            if dif[i] > 0 and dif[i] > dea[i] and dif[i-1] <= dea[i-1]:
                signal[i] = 1
            elif dif[i] < 0 and dif[i] < dea[i] and dif[i-1] >= dea[i-1]:
                signal[i] = -1
            elif dif[i] > dea[i]:
                signal[i] = 0.5
            else:
                signal[i] = -0.5
        return signal
    
    def calculate_macd_strength(self, dif: np.ndarray, dea: np.ndarray) -> float:
        """MACD 趋势强度"""
        if len(dif) < 2 or np.isnan(dif[-1]):
            return 0.0
        dif_value = dif[-1]
        dif_slope = dif[-1] - dif[-2] if len(dif) > 1 else 0
        return np.clip(dif_value / 10 + dif_slope / 2, -1, 1)

class RSICalculator:
    """RSI 计算器"""
    
    def __init__(self, period: int = 14):
        self.period = period
    
    def calculate(self, prices: np.ndarray) -> np.ndarray:
        """计算 RSI"""
        deltas = np.diff(prices)
        gains = np.where(deltas > 0, deltas, 0)
        losses = np.where(deltas < 0, -deltas, 0)
        result = np.full_like(prices, np.nan)
        avg_gain = np.mean(gains[:self.period])
        avg_loss = np.mean(losses[:self.period])
        for i in range(self.period, len(prices)):
            if i == self.period:
                rs = avg_gain / (avg_loss + 1e-10)
            else:
                avg_gain = (avg_gain * (self.period - 1) + gains[i]) / self.period
                avg_loss = (avg_loss * (self.period - 1) + losses[i]) / self.period
                rs = avg_gain / (avg_loss + 1e-10)
            result[i] = 100 - 100 / (1 + rs) if avg_loss > 0 else 100
        return result
    
    def calculate_rsi_signals(self, prices: np.ndarray) -> np.ndarray:
        """RSI 交易信号"""
        rsi = self.calculate(prices)
        signal = np.zeros_like(prices)
        for i in range(1, len(prices)):
            if np.isnan(rsi[i]) or np.isnan(rsi[i-1]):
                continue
            if rsi[i] < 30 and rsi[i] > rsi[i-1]:
                signal[i] = 1  # 超卖反弹
            elif rsi[i] > 70 and rsi[i] < rsi[i-1]:
                signal[i] = -1  # 超买回调
            elif rsi[i] > 50:
                signal[i] = 0.5
            else:
                signal[i] = -0.5
        return signal

class TrendResonanceStrategy:
    """趋势共振策略"""
    
    def __init__(self, 
                 ma_short: int = 10, ma_long: int = 30,
                 rsi_period: int = 14,
                 entry_threshold: float = 0.6,
                 exit_threshold: float = 0.3):
        self.ma = MovingAverageCalculator(ma_short, ma_long)
        self.macd = MACDCalculator()
        self.rsi = RSICalculator(rsi_period)
        self.entry_threshold = entry_threshold
        self.exit_threshold = exit_threshold
    
    def calculate_resonance_score(self, prices: np.ndarray) -> Tuple[float, Dict]:
        """计算共振评分"""
        ma_signal, _ = self.ma.calculate_ma_signals(prices)
        macd_signal = self.macd.calculate_macd_signals(prices)
        rsi_signal = self.rsi.calculate_rsi_signals(prices)
        dif, dea, _ = self.macd.calculate(prices)
        rsi = self.rsi.calculate(prices)
        ma_strength = self.ma.calculate_ma_strength(prices)
        macd_strength = self.macd.calculate_macd_strength(dif, dea)
        rsi_value = rsi[-1] / 100 if not np.isnan(rsi[-1]) else 0.5
        rsi_strength = 2 * (rsi_value - 0.5)
        
        total_signal = (ma_signal[-1] + macd_signal[-1] + rsi_signal[-1]) / 3
        total_strength = (ma_strength + macd_strength + rsi_strength) / 3
        
        indicators = {
            'ma_signal': ma_signal[-1],
            'macd_signal': macd_signal[-1],
            'rsi_signal': rsi_signal[-1],
            'ma_strength': ma_strength,
            'macd_strength': macd_strength,
            'rsi_strength': rsi_strength
        }
        
        return total_signal, total_strength, indicators
    
    def generate_signal(self, prices: np.ndarray, current_position: int = 0) -> TrendSignal:
        """生成交易信号"""
        score, strength, indicators = self.calculate_resonance_score(prices)
        
        if current_position == 0:
            if score >= self.entry_threshold:
                direction = 1
            elif score <= -self.entry_threshold:
                direction = -1
            else:
                direction = 0
        else:
            if abs(score) < self.exit_threshold:
                direction = 0
            else:
                direction = current_position
        
        confidence = min(1.0, abs(strength))
        
        return TrendSignal(
            timestamp=datetime.now(),
            direction=direction,
            strength=strength,
            indicators=indicators,
            confidence=confidence
        )

# 使用示例
if __name__ == '__main__':
    np.random.seed(42)
    dates = pd.date_range('2024-01-01', '2025-06-01', freq='D')
    n = len(dates)
    price = 100 + np.cumsum(np.random.randn(n) * 0.8)
    
    strategy = TrendResonanceStrategy(
        ma_short=10, ma_long=30,
        rsi_period=14,
        entry_threshold=0.5,
        exit_threshold=0.2
    )
    
    signals = []
    position = 0
    equity_curve = [100]
    
    for i in range(30, n):
        current_prices = price[:i+1]
        signal = strategy.generate_signal(current_prices, position)
        signals.append(signal)
        if signal.direction != position:
            if signal.direction == 0:
                position = 0
            elif position == 0:
                position = signal.direction
            else:
                position = signal.direction
        ret = 0.01 * signal.direction * (price[i] - price[i-1]) / price[i-1]
        equity_curve.append(equity_curve[-1] * (1 + ret))
    
    print("趋势共振量化策略回测结果")
    print(f"回测期间: {dates[0].date()} ~ {dates[-1].date()}")
    print(f"交易信号数: {len(signals)}")
    long_signals = sum(1 for s in signals if s.direction == 1)
    short_signals = sum(1 for s in signals if s.direction == -1)
    flat_signals = sum(1 for s in signals if s.direction == 0)
    print(f"做多次数: {long_signals}, 做空次数: {short_signals}, 空仓次数: {flat_signals}")
    print(f"最终净值: {equity_curve[-1]:.4f}")
    print(f"总收益率: {(equity_curve[-1] - 1) * 100:.2f}%")
```
### 数据来源
- Reddit: r/quant, r/algorithmictrading, r/investing
- 小红书: 趋势跟踪、量化策略
- 知乎: 量化交易、技术分析
- 其他: TradingView, Investopedia, AQR
